cmd/kubeadm/app/apis/kubeadm/v1beta3/types.go

```diff
@@ -0,0 +1,496 @@
+
+	// Etcd holds configuration for etcd.
+	Etcd Etcd `json:"etcd,omitempty"`
+
+	// Networking holds configuration for the networking topology of the cluster.
+	Networking Networking `json:"networking,omitempty"`
+
+	// KubernetesVersion is the target version of the control plane.
+	// Optional: Defaulted to latest stable 1.x.y release. Internet connection is required for the defaulting.
+	// +optional
+	KubernetesVersion string `json:"kubernetesVersion,omitempty"`
+
+	// ControlPlaneEndpoint sets a stable IP address or DNS name for the control plane; it
+	// can be a valid IP address or a RFC-1123 DNS subdomain, both with optional TCP port.
+	// In case the ControlPlaneEndpoint is not specified, the AdvertiseAddress + BindPort
+	// are used; in case the ControlPlaneEndpoint is specified but without a TCP port,
+	// the BindPort is used.
+	// Possible usages are:
+	// e.g. In a cluster with more than one control plane instances, this field should be
+	// assigned the address of the external load balancer in front of the
+	// control plane instances.
+	// e.g.  in environments with enforced node recycling, the ControlPlaneEndpoint
+	// could be used for assigning a stable DNS to the control plane.
+	// Optional: If unspecified, AdvertiseAddress + BindPort are used (see above)
+	// +optional
+	ControlPlaneEndpoint string `json:"controlPlaneEndpoint,omitempty"`
+
+	// APIServer contains extra settings for the API server control plane component
+	APIServer APIServer `json:"apiServer,omitempty"`
+
+	// ControllerManager contains extra settings for the controller manager control plane component
+	ControllerManager ControlPlaneComponent `json:"controllerManager,omitempty"`
+
+	// Scheduler contains extra settings for the scheduler control plane component
+	Scheduler ControlPlaneComponent `json:"scheduler,omitempty"`
+
+	// DNS defines the options for the DNS add-on installed in the cluster.
+	DNS DNS `json:"dns,omitempty"`
+
+	// CertificatesDir specifies where to store or look for all required certificates.
+	// Optional: Defaulted to "/etc/kubernetes/pki"
+	// +optional
+	CertificatesDir string `json:"certificatesDir,omitempty"`
+
+	// ImageRepository sets the container registry to pull images from.
+	// If empty, `k8s.gcr.io` will be used by default; in case of kubernetes version is a CI build (kubernetes version starts with `ci/` or `ci-cross/`)
+	// `gcr.io/kubernetes-ci-images` will be used as a default for control plane components and for kube-proxy, while `k8s.gcr.io`
+	// will be used for all the other images.
+	// Optional: Defaulted to "k8s.gcr.io" or "gcr.io/kubernetes-ci-images" (see above)
+	// +optional
+	ImageRepository string `json:"imageRepository,omitempty"`
+
+	// UseHyperKubeImage controls if hyperkube should be used for Kubernetes components instead of their respective separate images
+	// Optional: Defaulted to false
+	// +optional
+	UseHyperKubeImage bool `json:"useHyperKubeImage,omitempty"`
```

Comment:
should rely on v1beta3 to remove this ? ref https://github.com/kubernetes/kubeadm/issues/1889

we should think about it before shipping v1beta3, as we are able to remove fields only when bump the api version ?

cc @liggitt

Reply:
correct, fields should only be removed on a version boundary

Reply:
@liggitt @timothysc - two things come to my mind:

-  should we be able to round trip to `v1beta2` ? (If so, we could store this in an annotation or a field called `deprecatedUseHyperKubeImage` in order to avoid data loss)
-  are we expected to still support hyperkube if the release schedule does not match k/k ?

Reply:
> should we be able to round trip to v1beta2 ? (If so, we could store this in an annotation or a field called deprecatedUseHyperKubeImage in order to avoid data loss)

we've already done similar in previous iterations. @rosti can confirm.

> are we expected to still support hyperkube if the release schedule does not match k/k ?

we won't be able to, unless we allow the feed of custom hyperkube repository and image:tag, which will complicate upgrades.

Reply:
A custom repo can be achieved by the `imageRepository` field. The k8s version is more difficult. It would be best for the hyperkube image to continue to use the k8s versions.
As far as what needs to be done WRT hyperkube, I have already done some outline [on Slack](https://kubernetes.slack.com/archives/C13J86Z63/p1573206281062100?thread_ts=1573205684.061600&cid=C13J86Z63):

> Ok, so hyperkube is a drop-in replacement for the regular images. Therefore it should be relatively straight forward to just stop using it for the few users that do. It would probably cause problems for users with custom hyperkube images.
> Hence my vision is the following:
> - Remove the useHyperKubeImage field from v1beta3
> - Continue acknowledging and using the field when set to true in v1beta1 & v1beta2.
> - Print warnings of deprecation of hyperkube for now.
> - Drop support for hyperkube by ignoring it in ~1.20 or when v1beta1 & v1beta2 get deleted (whichever happens first).